packages/app/src/sandbox/worker/sw.ts

```diff
@@ -0,0 +1,210 @@
+}
+
+self.addEventListener('message', async event => {
+  if (typeof event.data !== 'object' || event.data.$channel !== CHANNEL_NAME) {
+    return;
+  }
+
+  const message = event.data as MessageSentToWorker;
+  switch (message.$type) {
+    case 'worker/init': {
+      const nextRelayPort = event.ports[0];
+
+      invariant(
+        relayPortPromise.state === 'pending',
+        'Failed to initialize relay: relay port promise already fulfilled from previous evaluation.'
+      );
+
+      /**
+       * @fixme Looks like upon shell restart, the worker is still running,
+       * so this promise is already resolved. Resolving it again with
+       * the correct message port does nothing, and it keeps pointing
+       * to the previous (incorrect) message port.
+       */
+      relayPortPromise.resolve(nextRelayPort);
+      debug('[SW]: relay port resolved', nextRelayPort);
+      break;
+    }
+
+    case 'worker/ping': {
+      // We are only interested in clients sending this.
+      if (!(event.source instanceof Client)) {
+        return;
+      }
+
+      const client = await self.clients.get(event.source.id);
+      if (client) {
+        const pong: IWorkerPongMessage = {
+          $channel: CHANNEL_NAME,
+          $type: 'worker/pong',
+        };
+
+        // Send back the pong message to keep the client/worker
+        // communication from becoming idle (i.e. terminated).
+        client.postMessage(pong);
+      }
+      break;
+    }
+  }
+});
+
+function getResponse(request: Request): DeferredPromise<IResponseData> {
+  const requestId = generateRandomId();
+  const requestPromise = new DeferredPromise<IResponseData>();
+
+  // Add some response timeout so the worker doesn't hang indefinitely,
+  // making it hard to know what went wrong.
+  const timeout = setTimeout(() => {
+    pendingRequests.delete(requestId);
+    requestPromise.reject(
+      new Error(
+        `Failed to handle ${request.method} ${request.url} request: no response received from the BroadcastChannel within timeout. There's likely an issue with the relay/worker communication.`
+      )
+    );
+  }, 5000);
+
+  const requestMessage: IPreviewRequestMessage = {
+    $channel: CHANNEL_NAME,
+    $type: 'preview/request',
+    id: requestId,
+    url: request.url,
+    method: request.method,
+  };
+
+  pendingRequests.set(requestId, requestPromise);
+
+  // The worker delegates request resolution to the relay, which,
+  // in turn, forwards these request messages to the main frame.
+  sendToRelay(requestMessage);
+
+  return requestPromise.finally(() => clearTimeout(timeout));
+}
+
+self.addEventListener('fetch', event => {
+  const req = event.request.clone();
+  const parsedUrl = new URL(req.url);
+
+  /**
+   * TODO: figure out a way to filter bundler requests and
+   * local requests. Now, it only supports `public`
+   */
+  if (
+    parsedUrl.origin !== self.location.origin ||
+    !parsedUrl.pathname.startsWith('/public')
```

Comment:
@DeMoorJasper and @CompuIves, one of the essential parts of the service worker is to intercept requests only made by the sandbox in the iframe and ignore all other requests, like static files by the bundler. 

However, I'm struggling to find an efficient way to ignore bundler requests from among the ones I want to catch. For example, the bundler requests files from `static/**/*` and from root, which makes it hard to check each file should be caught by the SW. For now, as a temporary solution, I'm considering all files coming from `/public` to belong to the sandbox and should be intercepted, while all others are ignored. 

Ideally, we should put all bundler static files inside another unique folder, like `__csb_static`, then we can easily ignore everything coming from this folder. But the webpack configuration is so overwhelming, and I don't know if it can cause side effects in regular sandboxes.

Reply:
Well, I've been trying to tweak the webpack to bundle everything inside `__csb_static`, but this seems a massive work, and I don't think we should do it right (too many files to change and check). Once we validate this Service Worker solution and after deprecating the legacy editor, I will be more confident about making these changes. So, for now, let's go for the temp solution (restrict requests coming from `public` folder) and evaluate it with beta testers.